lib/pbench/agent/results.py

```diff
@@ -102,100 +108,123 @@ def make_result_tb(self):
+                        "Error removing failed tarball MD5, {}", tarball_md5
                     )
-            sys.exit(1)
+            raise RuntimeError(f"Error in making tarball.md5, ({exc})") from exc
 
 
 class CopyResultTb:
+    """CopyResultTb - Copies tarball from agent to server with the help of RestAPI PUT request
+    """
+
     chunk_size = 4096
 
     def __init__(self, controller, tarball, config, logger):
+        """__init__ - Initializes some required attributes
+
+            Args
+                tarball -- path of tarball
+                tarball_md5 -- path of tarball's .md5
+                server_rest_url -- base uri
+                upload_url -- url where tarball needs to be send
+        """
         self.tarball = Path(tarball)
-        if not self.tarball.exists():
-            logger.error("tarball does not exist, '{}'", tarball)
-            sys.exit(1)
         self.tarball_md5 = Path(f"{tarball}.md5")
-        if not self.tarball_md5.exists():
-            logger.error("tarball's .md5 does not exist, '{}'", self.tarball_md5)
-            sys.exit(1)
+        self.check_tb_tbmd5()
         self.logger = logger
-        server_rest_url = config.results.get("server_rest_url")
+        server_rest_url = config.get("results", "server_rest_url")
         self.upload_url = f"{server_rest_url}/upload/ctrl/{controller}"
 
+    def check_tb_tbmd5(self):
+        """check_tb_tbmd5 - checks the existence of tb and tarball's .md5 file."""
+        if not self.tarball.exists():
+            raise FileNotFoundError(f"Tarball does not exist, '{self.tarball}'")
+        if not self.tarball_md5.exists():
+            raise FileNotFoundError(
+                f"Tarball's .md5 does not exist, '{self.tarball_md5}'"
+            )
+
     def read_in_chunks(self, file_object):
         data = file_object.read(self.chunk_size)
         while data:
             yield data
             data = file_object.read(self.chunk_size)
 
-    def copy_result_tb(self):
+    def copy_result_tb(self, token):
+        """copy_result_tb - copies tb from agent to given upload URL
+
+            Args
+                token -- token is an Authorization key which will help to authorise
+                    user on the other end.
+        """
         files = [f for f in self.tarball.parent.iterdir() if f.is_file()]
         file_count = len(files)
         if file_count != 2:
-            self.logger.error(
-                "(internal): unexpected file count, {}, associated with tarball, '{}'",
-                file_count,
-                self.tarball,
+            raise RuntimeError(
+                f"(internal): unexpected file count, {file_count}, associated with tarball, '{self.tarball}'"
             )
-            sys.exit(1)
 
         with open(self.tarball_md5, "r") as md5fp:
             md5sum = md5fp.read().split()[0]
         filename = secure_filename(str(self.tarball))
```

Comment:
This is a bit late in the process, but I suddenly realized that creating an MD5 file in this script is pointless as we never need it. We should just be creating local data (either an object for each tarball or a dict) that allows storing MD5 values locally in memory and then looking them up later for the actual `PUT` operation.

Reply:
@dbutenhof, what are you proposing?  We have the file right here, we could hash it now rather than opening an external file to get the value.  (Is there any value in hashing the file where the code currently does it, and then storing the result in memory instead of on disk?  Or, should I move that hashing code to here?)

Reply:
The point is that `make_md5sum` computes the MD5 of the new tarball and writes it to a file, which we then read, and discard. It'd make a lot more sense for `make_md5sum` to compute the MD5 and return it rather than creating the file which we don't need for any other purpose. Python is capable of remembering data without using the file system! ;-)

Reply:
Fixed.